### PR TITLE
[Java] Generate valid code if no Authentication implementations present

### PR DESCRIPTION
**Pull request by jonfreedman**
resurrects https://github.com/OpenAPITools/openapi-generator/pull/2861

Add hasBasicAuthMethods & hasTokenAuthMethods methods, use to decide which authentication code to use in standard Java and Java+RestTemplate templates.

### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) before.
- [x] Run the shell script(s) under `./bin/` (or Windows batch scripts under`.\bin\windows`) to update Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit, and these must match the expectations made by your contribution. You only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the code or mustache templates for a language (`{LANG}`) (e.g. php, ruby, python, etc).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.3.x`, `5.0.x`. Default: `master`.
- [x] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

@bbdouglas @sreeshas @jfiala @lukoyanov @cbornet @jeff9finger @karismann @Zomzog @lwlee2608 @bkabrda

### CODE REVIEW

**Comment by jonfreedman** [comment]
Can someone with permission please re-try the failed checks

**Comment by jonfreedman** [comment]
I cannot replicate the travis-ci failure on my local branch, this command generates valid code
```sh
mvn --quiet clean compile -f modules/openapi-generator-maven-plugin/examples/java-client.xml
```

File: CI/samples.ci/client/petstore/java/test-manual/jersey1/ApiClientTest.java [hunk 1]
```diff
@@ -102,14 +102,9 @@ public void testSelectHeaderContentType() {
...
 
-        Authentication auth = auths.get("api_key");
-        assertNotNull(auth);
-        assertTrue(auth instanceof ApiKeyAuth);
-        ApiKeyAuth apiKeyAuth = (ApiKeyAuth) auth;
-        assertEquals("header", apiKeyAuth.getLocation());
-        assertEquals("api_key", apiKeyAuth.getParamName());
-
-        auth = auths.get("petstore_auth");
+        assertNull(auths.get("api_key"));
```
Comment (wing328) [inline CI/samples.ci/client/petstore/java/test-manual/jersey1/ApiClientTest.java:105]: why does auths.get("api_key") return null?

I remember some endpoints in the petstore spec are defined to use "api_key" for authentication

Comment (jonfreedman) [reply CI/samples.ci/client/petstore/java/test-manual/jersey1/ApiClientTest.java:105]: As per https://github.com/OpenAPITools/openapi-generator/pull/2861#issuecomment-493825852 we now have hasTokenAuthMethods in DefaultGenerator which controls whether an ApiKeyAuth is registered in the authentications map.  Sounds like you're saying that is not working - any idea why not?

**Comment by wing328** [comment]
Sorry I should have caught it earlier. Please rename `hasTokenAuthMethods` to `hasApiKeyAuthMethods` instead. Thanks.

**Comment by jonfreedman** [comment]
> Sorry I should have caught it earlier. Please rename `hasTokenAuthMethods` to `hasApiKeyAuthMethods` instead. Thanks.

@wing328 done

**Comment by wing328** [comment]
@jonfreedman I think I've figured out the issue and filed https://github.com/OpenAPITools/openapi-generator/pull/6347 to add the helper methods to determine if certain authentication is required in the spec. (basically we should not determine if certain files should be included in the post-process operations)

I wonder if you can merge the latest master into this branch or file a new PR using the helper methods I added instead.

Thank you.

**Comment by jonfreedman** [comment]
@wing328 I'll merge now

**Comment by jonfreedman** [comment]
OK lets see what state this is in now, I must say this project is really a challenge to work with

**Comment by jonfreedman** [comment]
The bitrise tests passed but apparently it can't send status back to github...

**Comment by jonfreedman** [comment]
@wing328 I believe this is now "green" (https://app.bitrise.io/build/eeed859d13299e21)

**Comment by wing328** [comment]
> OK lets see what state this is in now, I must say this project is really a challenge to work with

@jonfreedman thanks for the feedback. I agreed there's a learning curve. If you've any suggestions to make it easier to contribute, please share your feedback with us.

(For this particular case, there's a fundamental issue (bug) with how we selectively add authentication -related files in post process operations to start with and thanks for PR so that we can uncover the issue and fix it accordingly)

**Pull request merged**